File: Symulator.java
import java.util.Random;
public class Zadanie_3 extends Thread {
    static class Traktor {

        static int POLE = 1;
        static int START = 2;
        static int PRACA = 3;
        static int KONIEC_PRACY = 4;
        static int AWARIA = 5;
        static int UZUPELNIJ = 1000;
        static int PUSTY_ZBIORNIK = 0;

        int numer;
        int nawozy;
        int stan;
        POLE p;
        Random rand;

        public Traktor(int numer, int nawozy, POLE p) {
            this.numer = numer;
            this.nawozy = nawozy;
            this.stan = PRACA;
            this.p = p;
            rand = new Random();
        }

    public void run(){
        while(true){
            if(stan==POLE){
                if(rand.nextInt(2)==1){
                    stan=START;
                    nawozy=UZUPELNIJ;
                    System.out.println("Siac nawozy jedzie, traktor "+numer);
                            stan=p.start(numer);
                }
                else{
                    System.out.println("Postoje sobie jeszcze troche");
                }
            }
            else if(stan==START){
                System.out.println("Wyjechal, traktor "+numer);
                stan=PRACA;
            }
            else if(stan==PRACA){
                nawozy-=rand.nextInt(500);
                System.out.println("Traktor "+numer+" na polu");
                if(nawozy<=PUSTY_ZBIORNIK){
                    stan=KONIEC_PRACY;
                }
                else try{
                    sleep(rand.nextInt(1000));
                }
                catch (Exception e){}
            }
            else if(stan==KONIEC_PRACY){
                System.out.println("Traktor wraca do domu "+numer+" ilosc nawozow w siewniku "+nawozy);
                stan=p.laduj();
                if(stan==KONIEC_PRACY){
                    nawozy-=rand.nextInt(500);
                    System.out.println("Reszta nawozow w siewniku:  "+nawozy);
                    if(nawozy<=0) stan=AWARIA;
                }
            }
            else if(stan==AWARIA){
                System.out.println("Awaria traktora "+numer);
                p.zmniejsz();
            }
        }
    }
    }
            static class POLE {
                static int POLE=1;
                static int START=2;
                static int PRACA=3;
                static int KONIEC_PRACY=4;
                static int AWARIA=5;
                int ilosc_siewnikow;
                int ilosc_zajetych;
                int ilosc_traktorow;
                POLE(int ilosc_siewnikow,int ilosc_traktorow){
                    this.ilosc_siewnikow=ilosc_siewnikow;
                    this.ilosc_traktorow=ilosc_traktorow;
                    this.ilosc_zajetych=0;
                }
                synchronized int start(int numer){
                    ilosc_zajetych--;
                    System.out.println("Siewnik podpiety do traktora. Mozna ruszac "+numer);
                    return START;
                }
                synchronized int laduj(){
                    try{
                        Thread.currentThread().sleep(1000);
                    }
                    catch(Exception ie){
                    }
                    if(ilosc_zajetych<ilosc_siewnikow){
                        ilosc_zajetych++;
                        System.out.println("Traktor sprawny, podpiety i gotowy, mozna jechac na pole "+ilosc_zajetych);
                        return POLE;
                    }
                    else
                    {return KONIEC_PRACY;}
                }
                synchronized void zmniejsz(){
                    ilosc_traktorow--;
                    System.out.println("AWARIA");
                    if(ilosc_traktorow==ilosc_siewnikow) System.out.println("Ilosc trkatorow jaka sama jak siewnikow");
                }
            }

    static class Glowna {
                static int ilosc_traktorow = 10;
                static int ilosc_siewnikow = 5;
                static POLE pole;

                public Glowna() {
                }

                public static void main(String[] args) {
                    pole = new POLE(ilosc_siewnikow, ilosc_traktorow);
                    for (int i = 0; i < ilosc_traktorow; i++)
                        new Traktor(i, 2000, pole).run();
                }
            }
    }

File: symulator.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include "mpi.h"
#define REZERWA 500
#define POLE 1
#define START 2
#define PRACA 3
#define KONIEC_PRACY 4
#define AWARIA 5
#define uzupelnij 5000
int nawozy = 8000;
int LADUJ=1, NIE_LADUJ=0;
int liczba_procesow;
int nr_procesu;
int ilosc_traktorow;
int nr_traktora;
int ilosc_pol=4;
int ilosc_zasianych_pol=0;
int tag=1;
int wyjedz[2];
int zaladuj[2];
MPI_Status mpi_status;
void Wyjedz(int nr_samolotu, int stan) 
{
wyjedz[0]=nr_traktora;
wyjedz[1]=stan;
MPI_Send(&wyjedz, 2, MPI_INT, 0, tag, MPI_COMM_WORLD);
sleep(1);
}

void Pole(int liczba_procesow){
int nr_traktora,status;
ilosc_traktorow = liczba_procesow - 1;
if(rand()%2==1){
printf("Jest ladna pogoda. Nie ma blota mozna siac\n");
}
else{
printf("Pada a na polu jest bloto. Nie jedz siac bo ugrzezniesz\n");
}
printf("Trzeba zasiac %d pol\n", ilosc_pol);
sleep(2);
while(ilosc_pol<=ilosc_traktorow){
MPI_Recv(&zaladuj,2,MPI_INT,MPI_ANY_SOURCE,tag,MPI_COMM_WORLD, &mpi_status);
nr_traktora=zaladuj[0];
status=zaladuj[1];
if(status==1){
printf("Traktor %d jest na polu\n", nr_traktora);
}
if(status==2){
printf("W siewniku sa nawozy traktor %d moze jechac na pole nr %d\n", nr_traktora, ilosc_zasianych_pol);
ilosc_zasianych_pol--;
}
if(status==3){
printf("Traktor %d sieje\n", nr_traktora);
}
if(status==4){
if(ilosc_zasianych_pol<ilosc_pol){
ilosc_zasianych_pol++;
MPI_Send(&LADUJ, 1, MPI_INT, nr_traktora, tag, MPI_COMM_WORLD);
}
else{
MPI_Send(&NIE_LADUJ, 1, MPI_INT, nr_traktora, tag, MPI_COMM_WORLD);
}
}

if(status==5){
ilosc_traktorow--;
printf("Ilosc traktorow %d\n", ilosc_traktorow);
}
} 
printf("Program zakonczyl dzialanie:)\n");
}
void Traktor(){
int stan,suma,i;
stan=PRACA; 
while(1){
if(stan==1){
if(rand()%2==1){
stan=START;
nawozy=uzupelnij;
printf("Wyjezdza, traktor %d\n",nr_procesu);
Wyjedz(nr_procesu,stan);
}
else{
Wyjedz(nr_procesu,stan);
}
}
else if(stan==2){
printf("Wyjechalem, traktor %d\n",nr_procesu);
stan=PRACA;
Wyjedz(nr_procesu,stan);
}
else if(stan==3){
nawozy-=rand()%500; 
if(nawozy<=REZERWA){
stan=KONIEC_PRACY;
printf("Traktor wraca uzupelnic zbiornik\n");
Wyjedz(nr_procesu,stan);
}

else{
for(i=0; rand()%10000;i++);
}
}
else if(stan==4){
int temp;
MPI_Recv(&temp, 1, MPI_INT, 0, tag, MPI_COMM_WORLD, &mpi_status);
if(temp==LADUJ){
stan=POLE;
printf("Powrot, traktor %d\n", nr_procesu);
}
else
{
nawozy-=rand()%500;
if(nawozy>0){
Wyjedz(nr_procesu,stan);
}
else{
stan=AWARIA;
printf("Uszkodzony uklad hydrauliczny\n");
Wyjedz(nr_procesu,stan);
return;
}
}
}
}
}
int main(int argc, char *argv[])
{
MPI_Init(&argc, &argv);
MPI_Comm_rank(MPI_COMM_WORLD,&nr_procesu);
MPI_Comm_size(MPI_COMM_WORLD,&liczba_procesow);
srand(time(NULL));
if(nr_procesu == 0)
Pole(liczba_procesow);
else 
Traktor();
MPI_Finalize();
return 0;
}
